Use `ConsolidateBuffer` for more aggressive per-worker pre-aggregation in `explode_one`

vmarcos opened the pull request:
This PR aims to improve the existing `explode_one` operator implementation in `src/timely-util/src/operator.rs` to more aggressively pre-aggregate on a per-worker basis when key domains are small. This goal is achieved by:

1. Employing a `ConsolidateBuffer` to call `consolidate_updates` across input batches as long as these input batches are of the same `time`.
2. Change the consolidation policy in `ConsolidateBuffer` to call `consolidate_updates` every time the output buffer doubles instead of only when it is filled with input tuples. In small key domains, this strategy limits as possible the lifetime of allocations and passes smaller buffers to `consolidate_updates`, attempting to mimic as much as possible in-place aggregation without introducing an additional data structure for this purpose.

### Motivation

  * This PR adds a known-desirable feature.

    Fixes #17271 

### Tips for reviewer

You can look at the individual commits for this PR. Adjustments to the `ConsolidateBuffer` typing were carried out to allow it to be employed with a `unary` operator. Note that we can instead implement the logic here with a raw Timely operator to avoid allocating a new output buffer for every new set of input batches; however, the changes to enable broader use of `ConsolidateBuffer` are probably useful in any case.

Measurements for this alternative have been collected in an [internal spreadsheet](https://docs.google.com/spreadsheets/d/1cVJEJ6ZwMG2KYpcl0TFSDJbH4iibybEE3igBEgHoosQ/edit#gid=0). The setup is referred to in PR #17281, which is a direct algorithmic change to `explode_one`. Only one of these two PRs should be merged, and I favor the present one. Note that the performance results are broadly comparable between the two PRs, indicating that (a) the better software organization here does not cost us much, if at all; (b) further going down the road of a raw Timely operator seems to be unnecessary at this point.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
N/A
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note): N/A

antiguru commented on src/timely-util/src/operator.rs:
```diff
@@ -411,29 +413,28 @@ where
+                            out.give(&time, item);
                         }
                     });
+                    out.flush();
```
`ConsolidateBuffer` flushes on drop, so there's no need to call `flush` explicitly.

vmarcos replied:
Indeed, I removed this to make the use in `explode_one` more idiomatic and to improve the calling code style even further, added a `give_iterator` function to `ConsolidateBuffer`.

antiguru commented on src/timely-util/src/buffer.rs:
```diff
@@ -81,11 +83,17 @@ where
+        // Limit, if possible, the lifetime of the allocations for data
+        // and consolidate smaller buffers if we're in the lucky case
+        // of a small domain for D
+        if self.buffer.len() >= 2 * self.previous_len {
```
Trying to untangle this, this seems to ensure that the buffer doesn't grow beyond its initial capacity because we're flushing on length > capacity/2, which means that 2*previous length <= capacity. Is this reasoning correct?

teskje replied:
Seems correct to me! If we didn't flush, that means that the length is still <= half the capacity, which means we can double it at least once more without exceeding the capacity.

vmarcos replied:
Upon thinking about this, the tightest argument I could formulate was exactly what @teskje is saying. I added a code comment documenting the reasoning for why the two conditions work together to maintain an invariant across calls to `give`.

antiguru reviewed (commented):
This seems fine! I left some comments inline, and I think the most important is to explain the invariants that `ConsolidateBuffer` needs to uphold. I can't access the spreadsheet linked, maybe you need to share it? Thanks!

teskje reviewed (approved):
LGTM!

vmarcos commented:
In addition to the review comments above, the latest commit also addresses the comment in https://github.com/MaterializeInc/materialize/pull/17345#discussion_r1089127834, thus improving the logic in `ConsolidateBuffer` that ensures enough capacity is reserved after a `give_vec` call on `flush`.

antiguru commented on src/timely-util/src/buffer.rs:
```diff
@@ -73,6 +73,18 @@ where
+    /// Provides an iterator of elements to the buffer
+    pub fn give_iterator<I: Iterator<Item = (D, T, R)>>(
+        &mut self,
+        cap: &CapabilityRef<T>,
```
You might have to rebase to avoid merge skew. I think this type was removed from Timely and is called `InputCapability` now.

vmarcos replied:
Thanks for the catch! I have changed this reference to `InputCapability` now as well, similarly to `give`.
